Add metrics to Neo4j Enterprise

chrisvest opened the pull request:
Please review the documentation parts, @nawroth

davidegrohmann committed:
Add Metrics Kernel Extension

chrisvest committed:
Add neo4j-metrics to the enterprise distribution

Also add an integration test for the metrics kernel extension.

chrisvest committed:
Git-ignore a couple of neo4j-home directories that are created during build

chrisvest committed:
First attempt at adding a page to the manual about the metrics plugin

nawroth committed:
Fixing metrics docs.

chrisvest committed:
Polishing the metrics documentation

MishaDemianenko commented on advanced/management/pom.xml:
```diff
@@ -8,7 +8,6 @@
     <relativePath>../..</relativePath>
   </parent>
 
-  <groupId>org.neo4j</groupId>
```
artefact should have group id, this line should be restored

chrisvest replied:
Why?

MishaDemianenko replied:
it's same reason as below

MishaDemianenko commented on enterprise/metrics/pom.xml:
```diff
@@ -0,0 +1,112 @@
+    <relativePath>../..</relativePath>
+  </parent>
+  
+  <artifactId>neo4j-metrics</artifactId>
```
where is group id?

chrisvest replied:
Inherited from the parent pom.

MishaDemianenko replied:
yes that i know it's really hard to read that in case of you pom is to big even if IDEA suggest to remove it.
The only field that can be removed here is version field because that is inherited and submodule will never gonna have version different from a parent. That is a common pattern to show that component is part of something bigger.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/MetricsExtension.java:
```diff
@@ -0,0 +1,108 @@
+        if ( prefix.equals( MetricsSettings.metricsPrefix.getDefaultValue() ) )
+        {
+            // If default name and in HA, try to figure out a nicer name
+            if ( config.getParams().containsKey( "ha.server_id" ) )
```
Can we use constant here instead of plain names?

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/MetricsSettings.java:
```diff
@@ -0,0 +1,83 @@
+    public static Setting<Boolean> neoPageCacheEnabled = setting( "metrics.neo4j.pagecache.enabled", Settings.BOOLEAN, neoEnabled );
+    public static Setting<Boolean> neoCountsEnabled = setting( "metrics.neo4j.counts.enabled", Settings.BOOLEAN, neoEnabled );
+    public static Setting<Boolean> neoNetworkEnabled = setting( "metrics.neo4j.network.enabled", Settings.BOOLEAN, neoEnabled );
+//    public static Setting<Boolean> neoLogEnabled = setting( "metrics.neo4j.log.enabled", Settings.BOOLEAN, neoEnabled );
```
should be removed instead of commented?

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/CsvOutput.java:
```diff
@@ -0,0 +1,95 @@
+
+            case split:
+            {
+                if ( !config.get( csvPath ).exists() )
```
can we extract config.get(csvPath) here to it's own variable?

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/CsvReporterSingle.java:
```diff
@@ -0,0 +1,379 @@
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger( CsvReporter.class );
+    private static final Charset UTF_8 = Charset.forName( "UTF-8" );
```
`StandardCharsets.UTF_8`

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/CsvReporterSingle.java:
```diff
@@ -0,0 +1,379 @@
+                if ( file.exists() )
+                {
+                    // Archive existing log
+                    String fileName = file.getName();
```
Is there is any reason why we choosing different rotation strategy here comparing with other log files?

chrisvest replied:
Who knows. This is how the imported code is doing it.

MishaDemianenko replied:
Then lets change it?)

chrisvest replied:
And it's not really rotating the file. Just archiving any file that might already exist by the given name, when the metrics extension starts up.

MishaDemianenko replied:
What is the principal difference? you can call that different rotation threshold.
Different renaming strategies still remains, those should be the same. There is no reason to have different name schemas and 'rotators' for different log files.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/CsvReporterSingle.java:
```diff
@@ -0,0 +1,379 @@
+                    }
+                }
+
+                out = new PrintWriter( new OutputStreamWriter( new FileOutputStream( file ), UTF_8 ) );
```
Since we already have file and encoding we can use `PrintWriter(String fileName, String csn)` in that case we can avoid stream creation

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/GraphiteOutput.java:
```diff
@@ -0,0 +1,72 @@
+        {
+            // Setup Graphite reporting
+
+            final Graphite graphite = new Graphite( new InetSocketAddress( config.get( graphiteServer ).getHost(),
```
`config.get( graphiteServer )` can be extracted to variable

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/NetworkMetrics.java:
```diff
@@ -0,0 +1,124 @@
+            // logBufferWrites was registered above - same monitor, remember?
+            monitors.addMonitorListener( networkTransactionReads, "logdeserializer" );
+
+            registry.register( name( "neo4j.network", "master_network_tx_writes" ), new Gauge<Long>()
```
how `tx_writes` corresponds with `BytesWritten`?

MishaDemianenko replied:
Metrics names should be extracted to constants

chrisvest replied:
I don't know if it's better to come up with new names for the exported fields, or if we should prefer to keep compatible with the metrics plugin from contrib.

chrisvest replied:
Any input on that, @cleishm @akollegger ?

MishaDemianenko replied:
it can be the case that we need some better names for those but i think it's separate from a fact that those metrics names should be collected into one place and defined as constants instead of being here and a bit there

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/NetworkMetrics.java:
```diff
@@ -0,0 +1,124 @@
+                }
+            } );
+
+            registry.register( name( "neo4j.network", "slave_network_tx_writes" ), new Gauge<Long>()
```
Same question as above

MishaDemianenko commented on pom.xml:
```diff
@@ -720,6 +720,21 @@
+      <dependency>
+        <groupId>com.codahale.metrics</groupId>
+        <artifactId>metrics-core</artifactId>
+        <version>3.0.2</version>
```
Looks like version of those 3 artifacts should be the same? can we extract it?

chrisvest committed:
Clean up the metrics code a little bit

chrisvest committed:
More metrics clean up

chrisvest committed:
Introduce a Counter abstraction and use it to make ByteCountsMetric thread-safe

chrisvest committed:
Clean up from constants in metrics

chrisvest committed:
Restore pom.xml groupId tags that Intellij thinks are redundant

chrisvest committed:
Mention the new metrics component in CHANGES.txt

nawroth commented:
Looks good. For the future: How about adding descriptions to `MetricsSettings` and generate docs from that as well? And at some point the examples could be generated in a similar way. [SettingsResourceBundle](https://github.com/neo4j/neo4j/blob/2.3/community/kernel/src/main/java/org/neo4j/graphdb/factory/SettingsResourceBundle.java) does half of the work needed for that already.

jexp commented:
Please add some descriptions of the recorded metrics and screenshots of the configurable visualization / output integrations.

chrisvest committed:
Straighten out the life cycling of the metrics extension

This makes the StoreMigrator* tests pass, because the metrics extension will now skip loading by throwing an exception (about missing dependencies) at the correct time in its life cycle

chrisvest committed:
Add description annotations on the MetricsSettings

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/Counter.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Abstraction for the concept of a thread-safe counter.
+ */
+public final class Counter
```
We already have `org.neo4j.helpers.Counter`. If this class is a substitution then old one should be removed or deprecated if it's part of public API.

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/Counter.java:
```diff
@@ -0,0 +1,79 @@
+     * Create a new Counter object, initialised to zero.
+     * @return A new Counter.
+     */
+    public static Counter create()
```
What's the gain of using factory method instead of plain constructor here?

chrisvest replied:
The potential to return alternative implementations.

MishaDemianenko replied:
Can we KISS it now? when we have only one?

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/Counter.java:
```diff
@@ -0,0 +1,79 @@
+     * Get the sum of all the increments and additions this counter has seen so far.
+     * @return A long value.
+     */
+    public long sum()
```
`sum` is not really good name here imho, would propose `value` instead

chrisvest replied:
It follows the terminology from j.u.c.LongAdder.

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/Counter.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * Reset the value of the counter back to zero.
+     */
+    public void reset()
```
looks like never used

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/Counter.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * Increment the value of the counter by 1.
+     */
+    public void increment()
```
never used

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/MetricsExtension.java:
```diff
@@ -0,0 +1,117 @@
+        life.init();
+    }
+
+    public void start() throws Throwable
```
`@Override` missing

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/MetricsExtension.java:
```diff
@@ -0,0 +1,117 @@
+        life.start();
+    }
+
+    public void stop() throws Throwable
```
`@Override` missing

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/CsvReporterSingle.java:
```diff
@@ -0,0 +1,413 @@
+            archiveExistingLogFile();
+        }
+
+        out = new PrintWriter( new OutputStreamWriter( new FileOutputStream( file ), StandardCharsets.UTF_8 ) );
```
constructor PrintWriter(String fileName, String csn) can be used instead

chrisvest replied:
Super annoying that there's no constructor that takes the Charset instance directly.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/CsvReporterSingle.java:
```diff
@@ -0,0 +1,413 @@
+    }
+
+    @Override
+    public void report( SortedMap<String,Gauge> gauges,
```
What will happen if `lifecycle.stop` will be invoked during report call?

chrisvest replied:
We write through a PrintWriter. It'll just raise a boolean flag internally, that we won't check.

MishaDemianenko replied:
About which flag we talking about here?
Lets imagine that we are in `report` method in Thread1 and somebody stop the server and that is in another thread `Thread2`. Now we have a possibility to have closed out in the middle of report call, as example of possible problems? or something preventing that?

chrisvest replied:
This flag: http://hg.openjdk.java.net/jdk7u/jdk7u/jdk/file/e228aaace9c9/src/share/classes/java/io/PrintWriter.java#l67
In other words, PrintWriter hides underlying IOExceptions.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/NetworkMetrics.java:
```diff
@@ -0,0 +1,130 @@
+     *
+     * HA: MasterClientXXX.class -> Transactions written to network for commit (writes)
+     *
+     * KERNEL: "logdeserializer"  -> Bytes read from network (read - updates for slaves, commits from slaves for master)
```
comment should be updated

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/NetworkMetrics.java:
```diff
@@ -0,0 +1,130 @@
+            monitors.addMonitorListener( masterNetworkStoreWrites, ToNetworkStoreWriter.class.getName(),
+                    "storeCopier" );
+            monitors.addMonitorListener( slaveNetworkTransactionWrites, MasterClient210.class.getName() );
+            // logBufferWrites was registered above - same monitor, remember?
```
looks like not valid comment

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/NetworkMetrics.java:
```diff
@@ -0,0 +1,130 @@
+                    "storeCopier" );
+            monitors.addMonitorListener( slaveNetworkTransactionWrites, MasterClient210.class.getName() );
+            // logBufferWrites was registered above - same monitor, remember?
+            monitors.addMonitorListener( networkTransactionReads, "logdeserializer" );
```
Can we give it more descriptive name?

chrisvest replied:
Can we give what a more descriptive name?

lutovich commented on pom.xml:
```diff
@@ -29,6 +29,7 @@
     <logback-classic.version>1.1.2</logback-classic.version>
     <generate-config-docs-phase>prepare-package</generate-config-docs-phase>
     <hsqldb.version>2.3.2</hsqldb.version>
+    <metrics.version>3.0.2</metrics.version>
```
Maybe we should use the latest version which is 3.1.0 ?

chrisvest committed:
Remove org.neo4j.concurrent.Counters

chrisvest committed:
Update com.codahale.metrics dependency to latest 3.1.2

chrisvest committed:
More metrics cleanup

Including removing a metric (logdeserializer) that had not been working at least since the JTA removal, but probably didn't work before that either.

chrisvest committed:
Add generated documentation for the MetricsSettings to the manual

chrisvest commented:
Replaced by #5419
